syntax = "proto3";
package server_services;

service ImporterService {
    rpc UploadCSV (FileUploadRequest) returns (FileUploadResponse);
    rpc UploadCSVChunks (stream FileUploadChunksRequest) returns (FileUploadChunksResponse);
}

message FileUploadRequest {
    bytes file = 1; // DTD file as bytes
    string file_mime = 2;
    string file_name = 3;
}

message FileUploadResponse {
    bool success = 1; 
}

message FileUploadChunksRequest {
    bytes data = 1; // The chunk of the file
    string file_name = 2; // File name (optional, sent in the first chunk or metadata)
}

message FileUploadChunksResponse {
    bool success = 1;
    string message = 2;
}

service GroupByService {
    rpc FilterXML (FilterRequest) returns (FilterResponse);
    rpc SearchXML (SearchRequest) returns (SearchResponse);
    rpc GroupXML (GroupRequest) returns (GroupResponse);
    rpc OrderXML (OrderRequest) returns (OrderResponse);
}

message FilterRequest {
    string file_name = 1;
    string xpath_query = 2;
}

message FilterResponse {
    string query_result = 1;
}

message SearchRequest {
    string file_name = 1;
    string search_term = 2;
}

message SearchResponse {
    repeated string matching_nodes = 1;
}

message GroupRequest {
    string file_name = 1;
    repeated string group_by_xpaths = 2;
}

message GroupResponse {
    map<string, int32> grouped_data = 1; 
}

message OrderRequest {
    string file_name = 1;
    string order_by_xpath = 2;
    bool ascending = 3;
}

message OrderResponse {
    repeated string ordered_nodes = 1;
}